lib/OpenQA/WebSockets/Controller/Auth.pm

```diff
@@ -0,0 +1,54 @@
+# Copyright (C) 2019 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+package OpenQA::WebSockets::Controller::Auth;
+use Mojo::Base 'Mojolicious::Controller';
+
+use OpenQA::Schema;
+use OpenQA::Utils 'log_debug';
+use Mojo::Util 'hmac_sha1_sum';
+
+sub check {
+    my $self = shift;
+
+    my $headers   = $self->req->headers;
+    my $key       = $headers->header('X-API-Key');
+    my $hash      = $headers->header('X-API-Hash');
+    my $timestamp = $headers->header('X-API-Microtime');
+    my $user;
+    log_debug($key ? "API key from client: *$key*" : "No API key from client.");
+
+    my $schema  = OpenQA::Schema->singleton;
+    my $api_key = $schema->resultset('ApiKeys')->find({key => $key});
+    if ($api_key) {
+        if (time - $timestamp <= 300) {
+            my $exp = $api_key->t_expiration;
+            # It has no expiration date or it's in the future
+            if (!$exp || $exp->epoch > time) {
+                if (my $secret = $api_key->secret) {
+                    my $sum = hmac_sha1_sum($self->req->url->to_string . $timestamp, $secret);
```

Comment:
no way to avoid duplicating this?

Reply:
I just moved the code here from the old `Mojolicious::Lite` app (no changes to the logic have been made). I do see some other uses of `hmac_sha1_sum` in `OpenQA::WebAPI`, but no copypasta. I guess it might make sense to look into sharing more code between WebAPI and WebSockets in a followup PR. If you spot more such cases please point them out with a comment.

Reply:
I see. github folding the deletes and showing the additions fooled me.